Remotion/Web/UnitTests/Core/Context/HttpContextStorageProviderNetCoreTest.cs

```diff
@@ -0,0 +1,112 @@
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Threading;
+using System.Web;
+using NUnit.Framework;
+using Remotion.Development.Web.UnitTesting.AspNetFramework;
+using Remotion.Web.Context;
+
+namespace Remotion.Web.UnitTests.Core.Context
+{
+#if !NETFRAMEWORK
+  [TestFixture]
+  public class HttpContextStorageProviderNetCoreTest
+  {
+    private HttpContextStorageProvider _provider;
+    private HttpContext _testContext;
+
+    [SetUp]
+    public void SetUp ()
+    {
+      _testContext = HttpContextHelper.CreateHttpContext("x", "y", "z");
+      HttpContext.Current = _testContext;
+      _provider = new HttpContextStorageProvider();
+    }
+
+    [Test]
+    public void DataRestoredAfterSafeContextBoundaryDisposed ()
+    {
+      _provider.SetData("test", 123);
+      using (_provider.OpenSafeContextBoundary())
+      {
+        var safeContextResult = _provider.GetData("test");
+        Assert.That(safeContextResult, Is.Null);
+      }
+
+      var result = _provider.GetData("test");
+      Assert.That(result, Is.EqualTo(123));
+    }
+
+    [Test]
+    public void SafeContextBoundaryDataUnavailableAfterDispose ()
+    {
+      using (_provider.OpenSafeContextBoundary())
+      {
+        _provider.SetData("test", 123);
+        var safeContextResult = _provider.GetData("test");
+        Assert.That(safeContextResult, Is.EqualTo(123));
+      }
+
+      var result = _provider.GetData("test");
+      Assert.That(result, Is.Null);
+    }
+
+    [Test]
+    public void SafeContextBoundaryCreationClearsData ()
+    {
+      _provider.SetData("test", 123);
+      _provider.OpenSafeContextBoundary();
+
+      Assert.That(HttpContext.Current.Items["Rm.DD"], Is.Null);
+    }
+
+    [Test]
+    public void HttpContextSwitchOnSameThreadResetsData ()
+    {
+      _provider.SetData("test", 123);
+
+      HttpContext.Current = HttpContextHelper.CreateHttpContext("a", "b", "c");
+
+      var result = _provider.GetData("test");
+      Assert.That(result, Is.Null);
+    }
+
+    [Test]
+    public void HttpContextSharedToNewThreadContainsSameData ()
+    {
+      _provider.SetData("test", 123);
+
+      var context = HttpContext.Current;
+      int? threadResult = null;
+
+      var t = new Thread(
+          ctx =>
+          {
+            HttpContext.Current = (HttpContext)ctx;
+            threadResult = (int)_provider.GetData("test");
+          });
+      t.Start(context);
```

Comment:
readability question: would it be easier to read to just capture `context` via a closure instead of passing it as an argument? For production code, this pattern is certainly the best way to go so closures are avoided, but in tests, it makes it harder to trace where things are coming from.